### PR TITLE
[Net] Split DNS resolving functionality out of net structures

### PR DESCRIPTION
**Pull request by Fuzzbawls**
Port of https://github.com/bitcoin/bitcoin/pull/7868

This brings CNetAddr/CSubNet/CService one step closer to being dumb storage structures. By forcing addresses to be resolved elsewhere, the implementation details are free to change. In particular, this is necessary for making the resolves fully async, which is necessary in a model in which the entire connection process is asynchronous.

### CODE REVIEW

File: src/netbase.cpp [hunk 1]
```diff
@@ -694,19 +696,19 @@ CNetAddr::CNetAddr(const struct in6_addr& ipv6Addr)
     SetRaw(NET_IPV6, (const uint8_t*)&ipv6Addr);
 }
 
-CNetAddr::CNetAddr(const char* pszIp, bool fAllowLookup)
+CNetAddr::CNetAddr(const char *pszIp)
```
Comment (random-zebra) [inline src/netbase.cpp]: nit: I think that the convention of placing asterisks right next to the variable type for pointers (i.e. `const char* pszIp`) is preferred almost everywhere in the codebase (e.g. in this very same diff, in netbase.h). We should strive for consistency.

File: src/netbase.cpp [hunk 2]
```diff
@@ -694,19 +696,19 @@ CNetAddr::CNetAddr(const struct in6_addr& ipv6Addr)
...
 {
     Init();
     std::vector<CNetAddr> vIP;
-    if (LookupHost(pszIp, vIP, 1, fAllowLookup))
+    if (LookupHost(pszIp, vIP, 1, false))
         *this = vIP[0];
 }
 
-CNetAddr::CNetAddr(const std::string& strIp, bool fAllowLookup)
+CNetAddr::CNetAddr(const std::string &strIp)
```
Comment (random-zebra) [inline src/netbase.cpp]: same as before for reference `&` types.

File: src/netbase.cpp (hunk 1 above)
Comment (Fuzzbawls) [reply src/netbase.cpp]: yeah, these changes are a result of the cherry-pick from upstream. I can go back through and edit each commit to NOT make this change.

This, of course, opens up a debate topic on what we want to use as "standard" and include in the developer notes.

Comment (random-zebra) [reply src/netbase.cpp]: Yeah, in C++ this is probably a debate as old as "tab vs spaces".
I think we should go with the most used convention in the codebase (without necessarily specify it in the contribution guidelines) and, in any case, at least keep consistency inside a single PR (e.g. it doesn't make sense to keep a function signature as `CNetAddr(const char* pszIp)` in the header, but change it in the *.cpp to make it `CNetAddr(const char *pszIp)`).

Comment (Fuzzbawls) [reply src/netbase.cpp]: updated to remove the reference side change

**Pull request merged**